_posts/2019-01-25-building-apps-with-plug-router.md

```diff
@@ -0,0 +1,354 @@
+
+Want to skip the reading and just look at a code?
+Head over to [elixirschool/router_example](https://github.com/elixirschool/router_example).
+
+## Getting started
+
+To get started we need to do a couple of things:
+
+1. Generate a new project with `mix new --sup`
+1. Add the `plug_cowboy` dependency to `mix.exs`
+1. Put our router is into the supervision tree of our application
+
+Without further delay, let's get this show on the road and generate our new project:
+
+```shell
+$ mix new router_example --sup
+
+* creating README.md
+* creating .formatter.exs
+* creating .gitignore
+* creating mix.exs
+* creating config
+* creating config/config.exs
+* creating lib
+* creating lib/router_example.ex
+* creating lib/router_example/application.ex
+* creating test
+* creating test/test_helper.exs
+* creating test/router_example_test.exs
+
+Your Mix project was created successfully.
+You can use "mix" to compile it, test it, and more:
+
+    cd router_example
+    mix test
+
+Run "mix help" for more commands.
+```
+
+Let's head into our new directory and open `mix.exs` in the editor of your choice.
+Here we'll make small a small change by adding the `plug_cowboy` dependency:
+
+```elixir
+defp deps do
+  [
+    {:plug_cowboy, "~> 2.0"}
+  ]
+end
+```
+
+With that change in place we can fetch our dependencies with `mix deps.get` and proceed.
+Though we have not yet created our router, let's get the supervisor setup out of the way.
+We'll need to open `lib/router_example/application.ex` next so we can update our supervisor's children.
+The `plug_cowboy` package makes this step easy with the included `Plug.Cowboy.child_spec/3` function.
+Let's update our application's `start/2` function:
+
+```elixir
+  def start(_type, _args) do
+    children = [
+      Plug.Cowboy.child_spec(scheme: :http, plug: RouterExample.Router, options: [port: 8080])
+    ]
+
+    opts = [strategy: :one_for_one, name: RouterExample.Supervisor]
+    Supervisor.start_link(children, opts)
+  end
+end
+```
+
+Our setup is complete!
+Before we can run our application we'll need create our `RouterExample.Router` module which we'll do that next.
+
+## Stubs
+
+As was mentioned in the introduction for the sake of keeping the tutorial short and focused, we'll be stubbing the functions that would otherwise retrieve data from our datastore and persist it there.
+We won't waste too much time on this file we just need two functions: one to give us a collection of results (stubbing a database lookup) and a second to take our parameters and write them to the terminal (instead of persisting them to a store).
+
+Let's create a new file `lib/router_example/stubs.ex` and copy the following code into it:
+
+```elixir
+defmodule RouterExample.Stubs do
+  def portfolio_entries do
+    for x <- 1..10, do: %{name: "Project #{x}", image: "https://picsum.photos/400/300/?random"}
+  end
+
+  def submit_contact(params) do
+    IO.inspect(params)
```

Comment:
Perhaps `IO.inspect(params, label: "Submitted contact")`?

When playing with an example application, I often IO.inspect things, and labeling this will make it less confusing to read the output.

Reply:
This is a great idea, we even have a post on using `IO.inspect/2` with the label option 👍

Reply:
Should we link to that blog post on label with a short note?